packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,178 @@
+};
+
+/**
+ * Creates a JSON-RPC middleware for handling queued requests. This middleware
+ * intercepts JSON-RPC requests, checks if they require queueing, and manages
+ * their execution based on the specified options.
+ *
+ * @param options - Configuration options.
+ * @param options.messenger - A controller messenger used for communication with various controllers.
+ * @param options.useRequestQueue - A function that determines if the request queue feature is enabled.
+ * @returns The JSON-RPC middleware that manages queued requests.
+ */
+export const createQueuedRequestMiddleware = ({
+  messenger,
+  useRequestQueue,
+}: {
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >;
+  useRequestQueue: () => boolean;
+}): JsonRpcMiddleware<JsonRpcParams, Json> => {
+  return createAsyncMiddleware(
+    async (
+      req: JsonRpcRequest & {
+        origin?: string;
+        networkClientId?: NetworkClientId;
+      },
+      res,
+      next,
+    ) => {
+      const { origin, networkClientId: networkClientIdForRequest } = req;
+
+      if (!origin) {
+        throw new Error("Request object is lacking an 'origin'");
+      }
+
+      if (!networkClientIdForRequest) {
+        throw new Error("Request object is lacking a 'networkClientId'");
+      }
+
+      // if the request queue feature is turned off, or this method is not a confirmation method
+      // do nothing
+      if (!useRequestQueue() || !isConfirmationMethod(req.method)) {
+        next();
+        return;
+      }
+
+      await messenger.call(
+        QueuedRequestControllerActionTypes.enqueueRequest,
+        async () => {
+          if (req.method === 'wallet_switchEthereumChain') {
+            return next();
+          }
+
+          const isBuiltIn = isNetworkType(networkClientIdForRequest);
+          let networkConfigurationForRequest;
+          if (isBuiltIn) {
+            const builtIn = BUILT_IN_NETWORKS[networkClientIdForRequest];
+            if (builtIn.chainId) {
+              // only the infura provided ones have chainid (rpc doesnt)
+              networkConfigurationForRequest = builtIn;
+            }
+          }
+
+          networkConfigurationForRequest ??= messenger.call(
+            'NetworkController:getNetworkClientById',
+            networkClientIdForRequest,
+          ).configuration;
+
+          const currentProviderConfig = messenger.call(
+            'NetworkController:getState',
+          ).providerConfig;
+
+          const currentChainId = currentProviderConfig.chainId;
+
+          // if the 'globally selected network' is already on the correct chain
```

Comment:
[total nit] just help comprehension
```suggestion
          // if the 'globally selected network' is already on the correct chain for the request currently being processed
```

Reply:
feel free to ignore this

Reply:
marged it
![image](https://github.com/MetaMask/core/assets/1787231/836614ba-7775-4837-b83f-5422143af3b3)